Add "Upgrading from Godot 4.6 to Godot 4.7" page (#11959)
godotengine/godot-docs: tutorials/migrating/upgrading_to_godot_4.7.rst

Behavior changes
----------------

Rendering
~~~~~~~~~

.. note::

    The ``LinearToSRGB`` visual shader no longer clamps to the range ``[0.0, 1.0]``
    when using the Mobile or Forward+ renderer (`GH-113956`_).

.. note::

    ``CanvasItem`` now avoids adding the antialiasing feather when drawing lines (`GH-105122`_).
    The feather made lines appear thicker than intended, projects that relied on this behavior
    will have to be updated to draw a thicker line width.

Physics
~~~~~~~

.. note::

    The default ``area_mask`` for ``AudioStreamPlayer`` was changed to ``0`` (disabled) (`GH-107679`_).
    If you use the ``audio_bus_override`` feature on ``Area2D`` or ``Area3D``, **and** you
    use the ``AudioStreamPlayer`` default ``area_mask`` (just layer ``1`` ticked), you will need
    to reset the mask to layer ``1`` — otherwise, the bus overrides will stop working.
    If the mask was set to anything except layer ``1``, it will continue to work as expected.

.. note::

    The distance ``D`` in planes when using the Jolt physics engine has been corrected to use

>     the right sign convention (`GH-118948`_).

Reviewer: ```suggestion
    When using Jolt Physics as the 3D physics engine, ``WorldBoundaryShape3D`` will now use the same
    convention as Godot when applying ``WorldBoundaryShape3D.plane.d``, resulting in the sign of the plane
    distance being interpreted in the opposite way compared to Godot 4.6 (`GH-118948`_). You will need to flip
    the sign yourself to get the same behavior as in Godot 4.6.
```

.. note::

>     The default ``SoftBody3D`` mass was changed from ``0`` to ``1`` in the Jolt physics engine (`GH-116041`_).

Reviewer: There's technically two breakages rolled into godotengine/godot#116041, so I'm splitting them up here.

```suggestion
.. note::

    When using Jolt Physics as the 3D physics engine, ``SoftBody3D`` will no longer default its mass to ``0``,
    which resulted in an automatically calculated weight of 1 kg per point, resulting in a very high total
    mass for the body. Now instead it will default to 1 kg for the entire ``SoftBody3D``, same as Godot
    Physics (`GH-116041`_).

.. note::

    When using Jolt Physics as the 3D physics engine, ``SoftBody3D`` will now apply
    ``SoftBody3D.linear_stiffness`` in a way that better matches Godot Physics, and in a way that's more
    appropriate in general (`GH-116041`_). This will affect every ``SoftBody3D`` instance in one way or
    another, meaning you will need to re-tweak properties like ``SoftBody3D.linear_stiffness`` and
    ``SoftBody3D.damping_coefficient`` to achieve your desired behavior.
```

.. note::

    The area's overlap signals are now emitted whenever a soft body enters/exits its shapes in the
    Jolt physics engine (`GH-114198`_). To work around this breaking change, configure your collision

>     layers/masks so that the relevant interactions between ``Area3D`` and ``SoftBody3D`` are ignored.

Reviewer: ```suggestion
    When using Jolt Physics as the 3D physics engine, ``Area3D`` will now report overlaps with ``SoftBody3D``
    from its various signals and methods (`GH-114198`_). To work around this breaking change, configure your
    collision layers/masks such that any undesirable interactions between ``Area3D`` and ``SoftBody3D`` are
    ignored.
```

**Control**
Property ``accessibility_live`` changes type from ``DisplayServer.AccessibilityLiveMode`` to ``AccessibilityServer.AccessibilityLiveMode``         |✔️|                 |❌|                  |❌|                  `GH-116839`_
**RichTextLabel**
Enum field ``ImageUpdateMask.UPDATE_WIDTH_IN_PERCENT`` renamed to ``ImageUpdateMask.UPDATE_WIDTH_UNIT``                                            |❌|                 |✔️|                  |❌|                  `GH-112617`_
Method ``add_image`` changes ``width`` parameter type from ``int`` to ``float``                                                                    |✔️|                 |✔️ with compat|      |✔️|                  `GH-112617`_
Method ``add_image`` changes ``height`` parameter type from ``int`` to ``float``                                                                   |✔️|                 |✔️ with compat|      |✔️|                  `GH-112617`_
Method ``add_image`` renames ``width_in_percent`` parameter to ``width_unit`` and changes type from ``bool`` to ``RichTextLabel.ImageUnit``        |✔️|                 |✔️ with compat|      |✔️ with compat|      `GH-112617`_

> Method ``add_image`` renames ``height_in_percent`` parameter to ``height_unit`` and changes type from ``bool`` to ``RichTextLabel.ImageUnit``      |✔️|                 |✔️ with compat|      |✔️ with compat|      `GH-112617`_

Reviewer 2: C# source compatibility has caveats, if a full list of args is used it will work as is, but if argument is used by name (e.g., `widthInPercent: true`), it won't. So not sure if it should be marked as `❌`, or have a footnote, or renaming argument automatically implies this.

Author: Because the compat method uses the name `widthInPercent`, it won't break users using named arguments.

Reviewer 2: At least for some cases it does not work (see https://github.com/godotengine/godot/issues/120148):

`AddImage(tex, 110, widthInPercent : true);` works in 4.6.3, but fails to build in 4.7.

Reviewer 2: Seems like it is caused by compat method not having default arguments (in the C# binding, arguments are set in GDScript binding).

Author: Ah, you are right, in the generated overload all parameters are required so this will break. I have updated the table to say `❌` in this case. Sorry for missing that.

Constant ``IMPORT_USE_NAMED_SKIN_BINDS`` moved to enum ``ImportFlags``                                                    |✔️|                 |✔️|                  |❌|                  `GH-115788`_
**EditorVCSInterface**
Method ``_commit`` adds new ``amend`` parameter                                                                           |❌|                 |❌|                  |❌|                  `GH-117968`_
========================================================================================================================  ===================  ====================  ====================  ============

Behavior changes
----------------

Rendering
~~~~~~~~~

.. note::

    The ``LinearToSRGB`` visual shader no longer clamps to the range ``[0.0, 1.0]``
    when using the Mobile or Forward+ renderer (`GH-113956`_).

.. note::

    ``CanvasItem`` now avoids adding the antialiasing feather when drawing lines (`GH-105122`_).
    The feather made lines appear thicker than intended, projects that relied on this behavior
    will have to be updated to draw a thicker line width.

Physics
~~~~~~~

.. note::

>     The default ``area_mask`` for ``AudioStreamPlayer`` was changed to ``0`` (disabled) (`GH-107679`_).

Reviewer 3: To `0` from what?

Author: The note mentions later the value used to be `1`. I added `from 1` to this line, is this better?

    which resulted in an automatically calculated weight of 1 kg per point, resulting in a very high total
    mass for the body. Now instead it will default to 1 kg for the entire ``SoftBody3D``, same as Godot
    Physics (`GH-116041`_).

.. note::

    When using Jolt Physics as the 3D physics engine, ``SoftBody3D`` will now apply
    ``SoftBody3D.linear_stiffness`` in a way that better matches Godot Physics, and in a way that's more
    appropriate in general (`GH-116041`_). This will affect every ``SoftBody3D`` instance in one way or
    another, meaning you will need to re-tweak properties like ``SoftBody3D.linear_stiffness`` and
    ``SoftBody3D.damping_coefficient`` to achieve your desired behavior.

.. note::

    When using Jolt Physics as the 3D physics engine, ``Area3D`` will now report overlaps with ``SoftBody3D``
    from its various signals and methods (`GH-114198`_). To work around this breaking change, configure your
    collision layers/masks such that any undesirable interactions between ``Area3D`` and ``SoftBody3D`` are
    ignored.

Input
~~~~~

.. note::

>     The device ID for mouse and keyboard has changed from ``0`` because some joypads may use it (`GH-116274`_).

Reviewer 3: From `0` to what?

Author: They were changed to the value of the new constants `DEVICE_ID_MOUSE` and `DEVICE_ID_KEYBOARD` which are mentioned later in the note. How would you rewrite the note to make it less confusing?

Reviewer 3: ```suggestion
    The device IDs for mouse and keyboard were changed from ``0`` to ``InputEvent.DEVICE_ID_MOUSE`` and ``InputEvent.DEVICE_ID_KEYBOARD`` because some joypads may use ``0`` as their ID (`GH-116274`_).
```

    Check the input event by type or compare the device ID ``InputEvent.device`` to the constants
    ``InputEvent.DEVICE_ID_MOUSE`` and ``InputEvent.DEVICE_ID_KEYBOARD`` instead.

GDScript
~~~~~~~~

.. note::

    Setting the element of packed arrays no longer calls the setter for the entire packed array property

>     (`GH-113228`_).

Reviewer 4: Please add a mention of [115763](https://github.com/godotengine/godot/pull/115763). Users must now return a value explicitly in methods that implicitly inherit a return type from a parent class.

Author: Added in f3cfaeb20. Feel free to suggest different content for the note, my explanation may not be accurate.

    another, meaning you will need to re-tweak properties like ``SoftBody3D.linear_stiffness`` and
    ``SoftBody3D.damping_coefficient`` to achieve your desired behavior.

.. note::

    When using Jolt Physics as the 3D physics engine, ``Area3D`` will now report overlaps with ``SoftBody3D``
    from its various signals and methods (`GH-114198`_). To work around this breaking change, configure your
    collision layers/masks such that any undesirable interactions between ``Area3D`` and ``SoftBody3D`` are
    ignored.

Input
~~~~~

.. note::

    The device ID for mouse and keyboard has changed from ``0`` because some joypads may use it (`GH-116274`_).
    Check the input event by type or compare the device ID ``InputEvent.device`` to the constants
    ``InputEvent.DEVICE_ID_MOUSE`` and ``InputEvent.DEVICE_ID_KEYBOARD`` instead.

GDScript
~~~~~~~~

.. note::

    Setting the element of packed arrays no longer calls the setter for the entire packed array property
    (`GH-113228`_).

.. note::

    Methods that inherit from a method with a typed return now inherit the return type as well,

>     requiring an explicit return statement in the override (`GH-115763`_`).

Reviewer 5: ```suggestion
    requiring an explicit return statement in the override (`GH-115763`_).
```

    ``SoftBody3D.damping_coefficient`` to achieve your desired behavior.

.. note::

    When using Jolt Physics as the 3D physics engine, ``Area3D`` will now report overlaps with ``SoftBody3D``
    from its various signals and methods (`GH-114198`_). To work around this breaking change, configure your
    collision layers/masks such that any undesirable interactions between ``Area3D`` and ``SoftBody3D`` are
    ignored.

Input
~~~~~

.. note::

    The device IDs for mouse and keyboard were changed from ``0`` to ``InputEvent.DEVICE_ID_MOUSE`` and
    ``InputEvent.DEVICE_ID_KEYBOARD`` because some joypads may use ``0`` as their ID (`GH-116274`_).
    Check the input event by type or compare the device ID ``InputEvent.device`` to the constants
    ``InputEvent.DEVICE_ID_MOUSE`` and ``InputEvent.DEVICE_ID_KEYBOARD`` instead.

GDScript
~~~~~~~~

.. note::

    Setting the element of packed arrays no longer calls the setter for the entire packed array property
    (`GH-113228`_).

.. note::

    Methods that inherit from a method with a typed return now inherit the return type as well,
    requiring an explicit return statement in the override (`GH-115763`_).

>     Add ``return null`` to the end of the method to fix the error.

Reviewer 6: I don't have time as of writing this, but I genuinely urge someone to come up with a better description based on the [RocketChat discussion](https://chat.godotengine.org/channel/gdscript?msg=eYxqZKskQdpHwQwC4) before this is accepted. It's bound to cause some serious confusion.

To make it somewhat brief, and I know it's not-so-difficult to think about for the more "technically-inclined", but from this description:
- It's difficult to infer that both exposed and user-written methods are affected by this change.
- It is especially relevant for **virtual methods** provided by the engine
- It's not clear what happens with overridden methods that return `void` (that is, nothing needs to be done)
	- Speaking of which, this mentions methods that "inherit" which not the engine's terminology. "Override" would likely be more appropriate and less confusing.
- How to fix a certain error is mentioned, but it's not directly explained what the error is.
